examples/wmt/train.py

```diff
@@ -0,0 +1,638 @@
+      'emb_dim': 1024,
+      'num_heads': 16,
+      'num_layers': 6,
+      'qkv_dim': 1024,
+      'mlp_dim': 4096,
+      'max_len': max(FLAGS.max_target_length, FLAGS.max_eval_target_length),
+      'share_embeddings': FLAGS.share_embeddings,
+      'logits_via_embedding': FLAGS.logits_via_embedding,
+  }
+
+  start_step = 0
+  rng = random.PRNGKey(FLAGS.random_seed)
+  rng, init_rng = random.split(rng)
+  input_shape = (FLAGS.batch_size, FLAGS.max_target_length)
+  target_shape = (FLAGS.batch_size, FLAGS.max_target_length)
+  model, cache_def = create_model(init_rng,
+                                  input_shape,
+                                  target_shape,
+                                  transformer_kwargs)
+  optimizer = create_optimizer(model,
+                               FLAGS.learning_rate,
+                               FLAGS.weight_decay)
+  # we access model only from optimizer below via optimizer.target
+  del model
+
+  if FLAGS.restore_checkpoints:
+    # Restore unreplicated optimizer + model state from last checkpoint.
+    optimizer = checkpoints.restore_checkpoint(FLAGS.model_dir, optimizer)
+    # Grab last step.
+    start_step = int(optimizer.state.step)
+
+  # Replicate optimizer.
+  optimizer = jax_utils.replicate(optimizer)
+
+  learning_rate_fn = create_learning_rate_scheduler(
+      base_learning_rate=FLAGS.learning_rate,
+      warmup_steps=FLAGS.warmup_steps)
+
+  p_train_step = jax.pmap(
+      functools.partial(
+          train_step,
+          learning_rate_fn=learning_rate_fn,
+          label_smoothing=FLAGS.label_smoothing,
+          use_bfloat16=FLAGS.use_bfloat16),
+      axis_name='batch')
+  p_eval_step = jax.pmap(
+      functools.partial(
+          eval_step,
+          label_smoothing=FLAGS.label_smoothing),
+      axis_name='batch')
+  p_pred_step = jax.pmap(
+      predict_step,
+      axis_name='batch',
+      static_broadcasted_argnums=(3, 4))  # eos token, max_length are constant
+
+  # We init the first set of dropout PRNG keys, but update it afterwards inside
+  # the main pmap'd training update for performance.
+  dropout_rngs = random.split(rng, n_devices)
+
+  metrics_all = []
+  t_loop_start = time.time()
+  for step, batch in zip(range(start_step, FLAGS.num_train_steps), train_iter):
+    # Shard data to devices and do a training step.
+    batch = common_utils.shard(jax.tree_map(lambda x: x._numpy(), batch))  # pylint: disable=protected-access
+    optimizer, metrics, dropout_rngs = p_train_step(
+        optimizer, batch, dropout_rng=dropout_rngs)
+    metrics_all.append(metrics)
+
+    # Save a Checkpoint
+    if step % FLAGS.checkpoint_freq == 0 and step > 0:
+      if jax.host_id() == 0 and FLAGS.save_checkpoints:
+        checkpoints.save_checkpoint(FLAGS.model_dir,
+                                    jax_utils.unreplicate(optimizer),
+                                    step)
+
+    # Periodic metric handling.
+    if step % FLAGS.eval_frequency == 0:
```

Comment:
Wow, that's a big if-statement! 😄 

Why not do `if step % FLAGS.eval_frequency != 0: continue` and save one indentation? I think it is quite difficult to parse whether the rest of the function is all under this if-statement or not.

(or factor out the whole eval part in a separate function)

Reply:
Sadly in the former 'if' statement all of those conditions are actually very necessary... I've reformated it into a single if.  

I've made the de-indentation change requested to ease reading.